src/util/small_map.h

```diff
@@ -0,0 +1,560 @@
+/*******************************************************************\
+
+Module: Small map
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Small map
+
+#ifndef CPROVER_UTIL_SMALL_MAP_H
+#define CPROVER_UTIL_SMALL_MAP_H
+
+#include <bitset>
+#include <cstdint>
+#include <cstring>
+#include <limits>
+#include <memory>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "invariant.h"
+
+// When we don't have constexpr
+#if !(defined(__GNUC__) || _MSC_VER >= 1900)
+
+template <std::size_t N>
+struct num_bitst
+{
+  static const std::size_t value = 1 + num_bitst<(N >> 1)>::value;
+};
+
+template <>
+struct num_bitst<1>
+{
+  static const std::size_t value = 1;
+};
+
+template <typename T, std::size_t B, typename U = std::integral_constant<T, 1>>
+struct indicator_maskt
+{
+  static const T value =
+    U::value |
+    indicator_maskt<T, B, std::integral_constant<T, (U::value << B)>>::value;
+};
+
+template <typename T, std::size_t B>
+struct indicator_maskt<T, B, std::integral_constant<T, 0>>
+{
+  static const T value = 0;
+};
+
+#endif
+
+/// Map from small integers to values
+///
+/// A data structure that maps small integers (in {0, ..., Num-1}) to values.
+/// It is designed to be more memory-efficient than std::map for this use case.
```

Comment:
Is this actually more memory efficient than arrays? The benchmark given in this comment for an empty `small_mapt<uint64t, uint32_t, 8>` - 32 bytes - is exactly as big as an `std::array<uint32_t, 4>`, with a full `small_mapt` being much worse. And arrays are also indexed by integers. Is the relation more favourable towards `small_mapt` with larger `Num`? If so, can you mention this here?